sympy/core/mul.py

```diff
@@ -195,6 +195,10 @@ def flatten(cls, seq):
                                 bargs.insert(0, ar)
                             bargs = [Add._from_args(bargs)]
                             rv = bargs, [], None
+            if a.is_commutative and b.is_commutative:
+                if ((a.is_finite == False and b.is_zero) or
+                        (b.is_finite == False and a.is_zero)):
+                    rv = [S.NaN], [], None
```

Comment:
Previously in the test you dropped it would have returned `z*i`.  So I just wondered about keeping that behaviour instead of explicitly returning NaN.  That is: we don't want it to collapse `z*i` to `0` but that does not imply it must return NaN...   (But keep in mind I did say I was not feeling confident about reviewing this!)